File: integracoes/meta-ads/scripts/contas.py
"""
contas.py — resolve cliente → conta de anúncios a partir de _memoria/contas-ads.md.
Fonte única de verdade para os scripts simples (diagnostico, auditoria, reels,
criativos, gerenciar). O token NÃO vive aqui (fica em credentials/, gitignored).
"""
import os

REPO_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", ".."))
CONTAS_PATH = os.path.join(REPO_ROOT, "_memoria", "contas-ads.md")

VALORES_VAZIOS = ("—", "", "(preencher)", "act_XXXXXXXXX")


class ContaError(Exception):
    pass


def _limpar(v):
    v = v.strip().strip("`").strip()
    return "" if v in VALORES_VAZIOS else v


def parse_contas(path=CONTAS_PATH):
    """Lê a tabela '## Contas Conectadas' e retorna lista de dicts."""
    if not os.path.exists(path):
        raise ContaError(f"Mapa de contas não encontrado: {path}")
    with open(path, "r", encoding="utf-8") as f:
        conteudo = f.read()

    contas = []
    in_secao = False
    headers = []
    for linha in conteudo.split("\n"):
        if "## Contas Conectadas" in linha:
            in_secao = True
            continue
        if in_secao and linha.startswith("## "):
            break
        if not in_secao or "|" not in linha or "---" in linha:
            continue
        partes = [p.strip() for p in linha.split("|") if p.strip()]
        if not partes:
            continue
        if not headers:
            headers = [h.lower().replace(" ", "_") for h in partes]
            continue
        row = {}
        for i, h in enumerate(headers):
            row[h] = _limpar(partes[i]) if i < len(partes) else ""
        contas.append(row)
    return contas


def resolver_cliente(path=CONTAS_PATH, nome=None):
    """Resolve um cliente. Sem nome e 1 só cliente com Meta → usa ele.
    Sem nome e vários → erro. Nome inexistente → erro."""
    contas = [c for c in parse_contas(path) if c.get("meta_ad_account")]
    if not contas:
        raise ContaError("Nenhum cliente com Meta Ad Account configurado em contas-ads.md.")
    if nome:
        alvo = nome.lower()
        match = [c for c in contas if alvo in c.get("cliente", "").lower()]
        if not match:
            disponiveis = ", ".join(c["cliente"] for c in contas)
            raise ContaError(f"Cliente '{nome}' não encontrado. Disponíveis: {disponiveis}")
        return match[0]
    if len(contas) > 1:
        disponiveis = ", ".join(c["cliente"] for c in contas)
        raise ContaError(f"Múltiplos clientes — use --cliente. Disponíveis: {disponiveis}")
    return contas[0]

File: server/src/contas.ts
import { readFile, writeFile } from "node:fs/promises";
import { join } from "node:path";

export type Conta = {
  cliente: string;
  metaAdAccount: string;
  handleIg: string;
  ativo: boolean;
};

export type ContaInput = {
  cliente: string;
  metaAdAccount?: string;
  igUserId?: string;
  handleIg?: string;
  googleAdsId?: string;
  ativo?: boolean;
};

// import.meta.dirname = server/src/ ; raiz do repo = dois níveis acima.
const REPO_ROOT = join(import.meta.dirname, "..", "..");
const CONTAS_PATH = join(REPO_ROOT, "_memoria", "contas-ads.md");

/** Faz parse da tabela markdown de contas. Recebe o conteúdo bruto do .md. */
export function parseContas(markdown: string): Conta[] {
  const out: Conta[] = [];
  for (const line of markdown.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed.startsWith("|")) continue;
    const cells = trimmed.split("|").slice(1, -1).map((c) => c.trim());
    if (cells.length < 6) continue;
    if (cells[0] === "Cliente") continue;        // cabeçalho
    if (/^-+$/.test(cells[0].replace(/\s/g, "-"))) continue; // separador ---
    if (cells[0] === "") continue;
    out.push({
      cliente: cells[0],
      metaAdAccount: cells[1],
      handleIg: cells[3],
      ativo: cells[5].toLowerCase() === "sim",
    });
  }
  return out;
}

/** Lê o arquivo real do repo e retorna as contas. Retorna [] se o arquivo não existir. */
export async function listContas(): Promise<Conta[]> {
  try {
    const md = await readFile(CONTAS_PATH, "utf-8");
    return parseContas(md);
  } catch (e) {
    if ((e as NodeJS.ErrnoException).code === "ENOENT") return [];
    throw e;
  }
}

const DEFAULT_CONTAS_MD = `# Contas de Anúncios — Mapa Cliente → Conta

Fonte única para resolução de conta nas integrações Meta/Google.
Token secreto NÃO vive aqui (fica em \`integracoes/credentials/\`, gitignored).

## Contas Conectadas

| Cliente | Meta Ad Account | IG User ID | Handle IG | Google Ads ID | Ativo |
|---------|-----------------|------------|-----------|---------------|-------|
`;

function rowCells(line: string): string[] | null {
  const trimmed = line.trim();
  if (!trimmed.startsWith("|")) return null;
  const cells = trimmed.split("|").slice(1, -1).map((c) => c.trim());
  if (cells.length < 6) return null;
  if (cells[0] === "Cliente") return null;
  if (/^-+$/.test(cells[0].replace(/\s/g, "-"))) return null;
  if (cells[0] === "") return null;
  return cells;
}

function buildRow(input: ContaInput, existing?: string[]): string {
  const metaAdAccount = input.metaAdAccount ?? existing?.[1] ?? "";
  const igUserId = input.igUserId ?? existing?.[2] ?? "";
  const handleIg = input.handleIg ?? existing?.[3] ?? "";
  const googleAdsId = input.googleAdsId ?? existing?.[4] ?? "";
  const ativo = input.ativo === undefined ? (existing?.[5] ?? "Sim") : input.ativo ? "Sim" : "Não";
  return `| ${input.cliente} | ${metaAdAccount} | ${igUserId} | ${handleIg} | ${googleAdsId} | ${ativo} |`;
}

/** Cadastra ou atualiza (por nome de cliente) uma linha em contas-ads.md. Retorna a lista atualizada. */
export async function saveConta(input: ContaInput): Promise<Conta[]> {
  let md: string;
  try {
    md = await readFile(CONTAS_PATH, "utf-8");
  } catch {
    md = DEFAULT_CONTAS_MD;
  }

  let found = false;
  const lines = md.split("\n").map((line) => {
    const cells = rowCells(line);
    if (cells && cells[0].toLowerCase() === input.cliente.toLowerCase()) {
      found = true;
      return buildRow(input, cells);
    }
    return line;
  });

  if (!found) {
    while (lines.length && lines[lines.length - 1].trim() === "") lines.pop();
    lines.push(buildRow(input));
  }

  const updated = lines.join("\n").trimEnd() + "\n";
  await writeFile(CONTAS_PATH, updated, "utf-8");
  return parseContas(updated);
}
